core/src/main/scala/org/ensime/core/AnalyzerManager.scala

```diff
@@ -0,0 +1,362 @@
+import org.ensime.config.RichEnsimeModule
+import org.ensime.util.FileUtils.toSourceFileInfo
+import org.ensime.util.ensimefile.richEnsimeFile
+import org.ensime.util.file.File
+
+import scala.collection.{ breakOut, mutable }
+
+trait ModuleFinder {
+  implicit val config: EnsimeConfig
+
+  def getModule(path: Path): Option[EnsimeProject] = {
+    config.projects.collectFirst {
+      case p if (p.sources.exists(f => path.toString.startsWith(f.toString))) =>
+        p
+    }
+  }
+
+  def getModule(fileInfo: SourceFileInfo): Option[EnsimeProject] = fileInfo match {
+    case SourceFileInfo(RawFile(f), _, _) => getModule(f)
+    case SourceFileInfo(ArchiveFile(f, _), _, _) => getModule(f)
+  }
+  def getModule(file: File): Option[EnsimeProject] = getModule(file.toPath)
+  def getModule(file: Either[File, SourceFileInfo]): Option[EnsimeProject] = file match {
+    case Left(f) => getModule(f)
+    case Right(sourceFileInfo) => getModule(sourceFileInfo)
+  }
+}
+
+sealed trait FileStatus {
+  def append(status: FileStatus): FileStatus
+}
+final case object NotLoaded extends FileStatus {
+  def append(status: FileStatus) = status match {
+    case Loaded => Loaded
+    case _ => NotLoaded
+  }
+}
+final case object Loaded extends FileStatus {
+  def append(status: FileStatus) = status match {
+    case Removed => Removed
+    case Unloaded => Unloaded
+    case _ => Loaded
+  }
+}
+final case object Removed extends FileStatus {
+  def append(status: FileStatus) = status match {
+    case Loaded => Loaded
+    case _ => Removed
+  }
+}
+final case object Unloaded extends FileStatus { // different from Removed
+  def append(status: FileStatus) = status match {
+    case Loaded => Loaded
+    case _ => Unloaded
+  }
+}
+
+final case class LoadedFilesData(statusOfFile: Map[SourceFileInfo, FileStatus], default: FileStatus)
+
+class AnalyzerManager(
+    broadcaster: ActorRef,
+    analyzerCreator: EnsimeProject => Props,
+    implicit val config: EnsimeConfig
+) extends Actor with ActorLogging with ModuleFinder with Stash {
+
+  // maps the active modules to their analyzers
+  private var analyzers: Map[EnsimeProject, ActorRef] = Map.empty
+  private val historyOfModule: mutable.Map[EnsimeProject, LoadedFilesData] = mutable.Map.empty
```

Comment:
This seems bit confusing for me. I would not mix usage of immutable/var and mutable/val, but rather pick one and stick with it.